### ISSUE TITLE
Proposal for better logging

### ISSUE THREAD

**Issue by Leooo** [emberjs/data#3874]
**TL;DR:
A proposal for a log to be used globally in Ember Objects, allowing to combine correct stack traces with a color code in the console, different log levels set globally or locally, and the ability to log all changes for a given model name / model instance that the user can set globally.**

Hello, I just wanted to share something I added recently to my app that i found incredibly useful to have better debugging tools for Ember-Data (I'm implementing an offline mode for ember-data with two stores, which started from a forked version of Ember-Sync, so have lots of work to do to understand model instance changes).

This would probably be a useful add-on for Ember, if only I had more experience and time I would try to create it but at my level this is not a possibility. It is still a work in progress.

Basically in Ember-Data, logging / stack traces are often not enough to understand the common problems such as state changes (including relationships) on a particular model instance. See https://github.com/emberjs/ember.js/issues/11863#issuecomment-135807116 for example. So previously I kept adding / removing console.logs to debug a particular object, taking hours to add all the logs until I get to the problem I'm trying to solve, then removing them again, only to add similar logs the next time i want to debug another model instance changes..

This is basically a mixin at the `Ember.Object` level, which determines which logs to show for this object. The log function:
- respects the correct stack traces in the console (= the line number is the real one).
- uses color codes in the console to improve readability.
- can be adjusted globally to send no logs in production (so this should not decrease the performance in production, though this would have to be checked as we are adding the mixing to every `Ember.Object`: this is maybe overkill).
- can be adjusted globally to show only the most important logs or a more detailed view (I'm using a color code for this: red=errors probably always shown, orange=possible dangers, green=normal logs, black=don't log except needed): `this.logLevel="green"`
- can be overriden in each object by overriding `logLevel`: `this.logLevel="black"` if we want more detailed logs on the object / class.
- moreover, we can choose globally to observe some model(s) (`modelNamesDebug`), or only some model instance(s) (`idsDebug`), if we want to understand what is going on with a specific model / model instance: so in one small change in the mixin's config and reloading the page we can log all future changes to this model instance(s) (it could even be possible to store all logs in memory, and display those of one object from the console ex-post - though this would decrease the app performance to store all the logs).

For this to work, the logs have to be written in this way:

``` javacript
//inside an Ember Object (controller, route, model etc.)

/*this will log if the global log level is above green (in the logMixin),
or if the local log level is above green (in the class where the log is applied)*/
this.log("green")("string describing the log",object1,..)

/*this will log if the global log level is above black (in the logMixin),
or if the local log level is above black (in the class where the log is applied),
or if modelNamesDebug.contains(modelName) (in the logMixin),
or if idsDebug.contains(id) (in the logMixin)*/
this.log("black",modelName,modelId)("string describing the log",object1,object2)
```

The whole setup is added below if some people are interested to copy, not very beautiful code for now but it works.

**1) logMixin (in mixins/log-mixin.js**

``` javascript
import Ember from 'ember';
import ENV from "frontend/config/environment";//replace by your own routine to detect your environment

//log line numbers
//http://stackoverflow.com/questions/13815640/a-proper-wrapper-for-console-log-with-correct-line-number
var Debugger = function() {
  var color=arguments.length ? arguments[0] : null;
  var klass=arguments.length>1 ? arguments[1] : null;
  var debug = {};
  var st=klass.constructor.toString();
  for (var i=2;i<arguments.length;i++) {
    st+=" "+arguments[i];
  }
  if (color) {
    for (var m in console)
      if (typeof console[m] == 'function')
        ////http://stackoverflow.com/questions/23381423/append-arguments-to-javascript-functions-rather-than-prepend-function-prototyp
        //wont preserve the correct line number (will take the one from the mixin given we work on the args)
        //this.debug[m] = console[m].bindAppendLV(window.console, klass/*.toString()*/);
        debug[m] = console[m].bind(window.console, '%c['+st+']',
            'color: '+color+';'+(color==="black" ? '' : ' font-weight:bold;')/*,klass*/);
  }
  else{
    for (var m in console)
      if (typeof console[m] == 'function')
        debug[m] = function(){};
  }
  return debug;
}

var LOG_LEVELS={all: [true,"black","blue","green","orange","red"], black: ["black","blue","green","orange","red"], 
  blue: ["blue","green","orange","red"], green: ["green","orange","red"], orange: ["orange","red"], red: ["red"]};

export default Ember.Mixin.create({
  log: function(color,modelName,id){//DO NOT use any of the final arguments with call or apply (only with bind)
    var gDebug=ENV.dotenv.environment==="development" ? "green" : "orange";//true,"black","blue","green","orange","red"
    var lDebug=this.logLevel ? this.logLevel : false;//true,"blue","green","orange","red"
    var modelNamesDebug=["profile","feedback"];//highlight / force log those models
    var idsDebug=[];//highlight / force log only the models with this id
    var onlyId=false;//if idsDebug is presen, debug only this id
    //var stringFetch=["feedback"];//enter an array of strings. If they are present on the log strings, they will be displayed
    color = color || "black";
    modelName = modelName || null;
    id = id || null;
    //minimum log level is set by gDebug
    //pages log levels can be more restrictive only than gDebug using lDebug
    //var debug= LOG_LEVELS["all"].indexOf(lDebug) > LOG_LEVELS["all"].indexOf(gDebug) ? lDebug : gDebug;
    //pages log levels are lDebug if exists, otherwise gDebug
    var debug= lDebug || gDebug;
    if (modelName && modelNamesDebug.contains(modelName)) {
      if (idsDebug.length) {
        if (id &&(idsDebug.contains(parseInt(id)) || idsDebug.contains(String(id)))) {debug =true;color="magenta";}
        else {if (!onlyId) {debug =true;color="violet";}}
      }
      else {debug =true;color="violet";}
    }
    if (debug!==true) {
      if (["black","blue","green","orange","red"].contains(color)) {
        if (!LOG_LEVELS[debug].contains(color)) {return function(){};}
      }
      else {return function(){};}
    }
    return Debugger(color, this,modelName,id).log;
  },
  /*log: function(){
    return this.logByColor("black")(arguments);
  },*/
});
```

**2) override Ember.Object in initializers/override-ember-object**

``` javascript
import Ember from 'ember';
import LogMixin from "frontend/mixins/log-mixin";

export default {
  name: 'override-ember-object',
  initialize: function() {
    Ember.Object.reopen(LogMixin,{});
  }
}
```

**3) use in controllers, models, routes, mixins.. or any Ember.Object**

``` javascript
RecordForSynchronization.reopen({
  //override the logLevel only if you want a specific one for this class:
  //logLevel: "green",
      toEmberData: function() {
        var self = this,
            record;
        self.log("green")("emberSync toEmberData");
        return new Ember.RSVP.Promise(function(resolve, reject) {
          record = self.createRecordInStore();
          self.setRelationships(record).then(function(rec){
            self.log("black",rec.constructor.modelName,rec.get("id"))("emberSync did set relationships",
              rec.constructor.modelName,rec.get("id"),rec);
            resolve(rec);
          },function(err){
            self.log("red")("error in recordforsynchronization promise",err,record,self);
          });
        });
      },
});
```

**Comment by Leooo** [comment]
Forgot to say, the link with ember-data is: imagine that in ember-inspector, you could click on one model name, and then see in the next requests all the operations involving instances of this model (serialisations, normalisations, fetching, state changes etc.). It would help a lot to understand everything. Just an Idea.

**Comment by sandstrom** [comment]
I would do it the other way around. Make a centralized logger without dependencies on Ember. Implement some type of pub/sub. Then add a simple shim on top of `Ember.Logger`.

```js
var myCentralizedLogger = { /* …not dependent on Ember… */ };

Ember.Logger = { 
  debug: function(msg, meta) { myCentralizedLogger.debug(msg, meta); },
  info: function(msg, meta) { myCentralizedLogger.info(msg, meta); },
  // ...
};

```

You can then subscribe to your centralized logger in various ways. Print to console, send error level to bug-reporting tool, send to your own backend, keep in local storage on the client, for later, etc.

What's currently in Ember make this very easy, whilst being flexible for different setups.

**Comment by Leooo** [comment]
That would be more the Ember current way yes, though I had to struggle a bit (I'm not a specialist) to be able to mix color codes, meta (to set properties used for conditional log levels such as modelName ad id) and correct line numbers (see the special form of the logging function `this.log(meta)(messages)`), so I don't think Ember.logger would allow that in its current form.

Such a logger would seem like a very special case, and probably most people are happy with the current Ember.logger framework. I just thought that, given that close to all ember-data code is about doing something on a modelName + id, systematically logging conditionally such information for debugging purposes (without adding logs everywhere though, so probably using the existing framework for promise descriptions and Ember-inspector) would bring a LOT to the framework. This is definitely not the priority I suppose and I am very happy with all the information I'm getting now with this logger which makes it immediate to observe state changes in ED objects for specific model instances / relationships.

**Comment by wecc** [comment]
I would be happy to see an [RFC](https://github.com/emberjs/rfcs) about this.

**Comment by stefanpenner** [comment]
closing in-favor of RFC

**Comment by rmharrison** [comment]
@stefanpenner: Following purposefully not including `Ember.Logger` in the modules API (https://github.com/emberjs/rfcs/pull/176#issuecomment-272566327, https://github.com/emberjs/ember.js/issues/15724), was there ever an RFC on how handle logging?

Was unable to find in https://github.com/emberjs/rfcs.
